smartclient/relayhandler/relayhandler.go

```diff
@@ -0,0 +1,322 @@
+	// a list of valid endpoints for the session and the RPCType used by the client
+	endpointSelectionStrategy EndpointSelectionStrategy
+
+	// Signer is the signer used to sign the relay request
+	signer Signer
+}
+
+func NewRelayHandler(
+	listenAddr string,
+	applicationQueryClient applicationTypes.QueryClient,
+	sessionQueryClient sessionTypes.QueryClient,
+	accountQueryClient authTypes.QueryClient,
+	blockQueryClient *client.BlocksQueryClient,
+	applicationAddress string,
+	endpointSelectionStrategy EndpointSelectionStrategy,
+	signer Signer,
+) *RelayHandler {
+	return &RelayHandler{
+		listenAddr:                listenAddr,
+		applicationQueryClient:    applicationQueryClient,
+		sessionQueryClient:        sessionQueryClient,
+		accountQueryClient:        accountQueryClient,
+		blockQueryClient:          blockQueryClient,
+		applicationAddress:        applicationAddress,
+		currentSessions:           make(map[string]*sessionTypes.Session),
+		servicesSessions:          make(map[string]utils.Observable[*sessionTypes.Session]),
+		endpointSelectionStrategy: endpointSelectionStrategy,
+		signer:                    signer,
+	}
+}
+
+func (relayHandler *RelayHandler) Start(ctx context.Context) error {
+	relayHandler.ctx = ctx
+
+	// get services supported by the application
+	applicationRequest := &applicationTypes.QueryGetApplicationRequest{
+		Address: relayHandler.applicationAddress,
+	}
+	appInfo, err := relayHandler.applicationQueryClient.Application(ctx, applicationRequest)
+	if err != nil {
+		return err
+	}
+
+	// create a session notifier for each service
+	// fetches the initial session info for them so the relay handler can start serving relays
+	servicesActiveSession := map[string]*sessionNotifier{}
+	for _, service := range appInfo.Application.Services {
+		sessionsNotifee, sessionsNotifier := utils.NewControlledObservable[*sessionTypes.Session](nil)
+		relayHandler.servicesSessions[service.Id] = sessionsNotifee
+
+		sessionInfo, err := relayHandler.fetchCurrentSession(ctx, service.Id)
+		if err != nil {
+			log.Printf("could not create a session notifier for service %s: %v", service.Id, err)
+			continue
+		}
+
+		servicesActiveSession[service.Id] = &sessionNotifier{
+			Notifier: sessionsNotifier,
+			Session:  sessionInfo,
+		}
+
+		relayHandler.currentSessions[service.Id] = sessionInfo
+	}
+
+	go relayHandler.providedServicesSessionsListener(ctx, servicesActiveSession)
+
+	// start listening for relays, this is a regular http server that uses ServeHTTP
+	// implemented by the relayHandler to handle http/ws relays
+	go http.ListenAndServe(relayHandler.listenAddr, relayHandler)
+
+	return nil
+}
+
+// providedServicesSessionsListener listens for new blocks and updates the session info for each service
+// if session last block is past the current block, it notifies the corresponding session notifier
+func (relayHandler *RelayHandler) providedServicesSessionsListener(
```

Comment:
I think the relay handler's session management business logic should be abstracted out into a `SessionCache`. Wdyt?

Reply:
Yes, it will greatly simplify code that is using it. A first attempt was to extract some of the logic into `session_connection.go`.  But itself should be written in terms of `SessionCache`